include/partition/edge_based_graph_reader.hpp

```diff
@@ -125,55 +128,51 @@ std::vector<OutputEdgeT> prepareEdgesForUsageInGraph(std::vector<extractor::Edge
     return graph_edges;
 }
 
-struct EdgeBasedGraphReader
+std::vector<extractor::EdgeBasedEdge> graphToEdges(const DynamicEdgeBasedGraph &edge_based_graph)
 {
-    EdgeBasedGraphReader(storage::io::FileReader &reader)
-    {
-        // Reads:  | Fingerprint | #e | max_eid | edges |
-        // - uint64: number of edges
-        // - EdgeID: max edge id
-        // - extractor::EdgeBasedEdge edges
-        //
-        // Gets written in Extractor::WriteEdgeBasedGraph
-
-        const auto num_edges = reader.ReadElementCount64();
-        const auto max_edge_id = reader.ReadOne<EdgeID>();
-
-        num_nodes = max_edge_id + 1;
-
-        edges.resize(num_edges);
-        reader.ReadInto(edges);
-    }
-
-    // FIXME: wrapped in unique_ptr since dynamic_graph is not move-able
-
-    std::unique_ptr<DynamicEdgeBasedGraph> BuildEdgeBasedGraph()
-    {
-        // FIXME: The following is a rough adaption from:
-        // - adaptToContractorInput
-        // - GraphContractor::GraphContractor
-        // and should really be abstracted over.
-        // FIXME: edges passed as a const reference, can be changed pass-by-value if can be moved
-
-        auto directed = splitBidirectionalEdges(edges);
-        auto tidied = prepareEdgesForUsageInGraph<DynamicEdgeBasedGraphEdge>(std::move(directed));
-
-        return std::make_unique<DynamicEdgeBasedGraph>(num_nodes, std::move(tidied));
-    }
+    NodeID max_turn_id = 0;
+    auto range = tbb::blocked_range<NodeID>(0, edge_based_graph.GetNumberOfNodes());
+    tbb::parallel_for(range, [&](const auto range) {
+        for (auto node = range.begin(); node < range.end(); ++node)
+        {
+            for (auto edge : edge_based_graph.GetAdjacentEdgeRange(node))
+            {
+                const auto &data = edge_based_graph.GetEdgeData(edge);
+                max_turn_id = std::max(max_turn_id, data.turn_id);
```

Comment:
Here can be a data race. It can fixed by 
* `max_turn_id` is atomic (bad)
* use thread local values 
* use `tbb::parallel_reduce`